### PR TITLE
Allow SVI to skip invalid updates

### LINKED ISSUE #953: Unhelpful Error Dirichlet Process Mixture Model
Following my post here, https://forum.pyro.ai/t/dirichlet-process-gaussian-mixture-model-typeerror-for-small-concentration-parameter/2685, when one tries creating a Dirichlet Process Mixture Model with 0 mixture components, the error message returned is 

`TypeError: iota shape must have every element be nonnegative, got (-1,).`

This isn't helpful in identifying the source of the error. Could a better message be returned?

### LINKED ISSUE #956: Skip current update in SVI if loss gets nan
cc @martinjankowiak

### PR DESCRIPTION
**Pull request by fehiepsi**
Resolves #956 and resolves #953. This feature should be disabled by default so that nan issue can put some concern for users.

### CODE REVIEW

File: numpyro/optim.py [hunk 1]
```diff
@@ -77,6 +77,24 @@ def eval_and_update(self, fn: Callable, state: _IterOptState) -> _IterOptState:
         out, grads = value_and_grad(fn)(params)
         return out, self.update(grads, state)
 
+    def eval_and_stable_update(self, fn: Callable, state: _IterOptState) -> _IterOptState:
+        """
+        Likes :meth:`eval_and_update` but when the value of objective function
```
Comment (martinjankowiak) [inline numpyro/optim.py:82]: typo: Like
nit: the objective function

File: numpyro/infer/svi.py [hunk 2]
```diff
@@ -146,7 +146,25 @@ def update(self, svi_state, *args, **kwargs):
         loss_val, optim_state = self.optim.eval_and_update(loss_fn, svi_state.optim_state)
         return SVIState(optim_state, rng_key), loss_val
 
-    def run(self, rng_key, num_steps, *args, progress_bar=True, **kwargs):
+    def stable_update(self, svi_state, *args, **kwargs):
+        """
+        Similar to :meth:`update` but returning the current state if the
```
Comment (martinjankowiak) [inline numpyro/infer/svi.py:151]: nit: Similar to :meth:`update` but returns the current state if the
        the loss or the new state contains invalid values.

File: test/infer/test_svi.py [hunk 3]
```diff
@@ -186,3 +186,19 @@ def guide():
...
+        var = numpyro.sample("var", dist.Exponential(1))
+        numpyro.sample("obs", dist.Normal(0, jnp.sqrt(var)), obs=0.)
+
+    def guide():
+        loc = numpyro.param("loc", 0.)
+        numpyro.sample("var", dist.Normal(loc, 10))
+
+    svi = SVI(model, guide, optim.Adam(1), Trace_ELBO())
+    svi_result = svi.run(random.PRNGKey(0), 1000, stable_update=stable_run)
+    assert jnp.isfinite(svi_result.params["loc"]) == stable_run
```
Comment (martinjankowiak) [inline test/infer/test_svi.py:204]: 👍

**Review by martinjankowiak — APPROVED** [review]

**Comment by fehiepsi** [comment]
Thanks for reviewing, @martinjankowiak!

**Review by martinjankowiak — APPROVED** [review]

**Pull request merged**